lib/recursive-methods.rb

```diff
@@ -1,41 +1,76 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  elsif n <= 1
+    return  1
+  else 
+    n * factorial(n-1)
+  end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n^2)
+# Space complexity: o(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    # raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s
+  else 
+    return s[-1] + reverse(s[0..-2]) 
```

Comment:
`s[1..-1]` creates a new array and  copies all the individual elements over and so is O(n) by itself.